File: 0695-max-area-of-island/0695-max-area-of-island.cpp
class Solution {
private:
    //                          up,      right,   down,    left
    vector<vector<int>> dirs = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
    int bfs(int r, int c, vector<vector<bool>>& vis, vector<vector<int>>& grid, int n, int m) {
        int cnt = 0;
        queue<pair<int, int>> q;
        vis[r][c] = true;
        q.push({r, c});
        while (!q.empty()) {
            int R = q.front().first;
            int C = q.front().second;
            q.pop();
            cnt++;
            for (int i = 0; i < dirs.size(); ++i) {
                int nr = R + dirs[i][0];
                int nc = C + dirs[i][1];
                if (nr >= 0 and nr < n and nc >= 0 and nc < m and
                    grid[nr][nc] and !vis[nr][nc]) {
                    q.push({nr, nc});
                    vis[nr][nc] = true;
                }
            }
        }
        return cnt;
    }

public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int n = grid.size();
        int m = grid[0].size();
        int maxArea = 0;
        vector<vector<bool>> vis(n, vector<bool>(m, false));
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                if (grid[i][j] and !vis[i][j]) {
                    maxArea = fmax(maxArea, bfs(i, j, vis, grid, n, m));
                }
            }
        }
        return maxArea;
    }
};

File: 0695-max-area-of-island/0695-max-area-of-island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        dirs = [[-1, 0], [0, 1], [1, 0], [0, -1]]
        n = len(grid)
        m = len(grid[0])
        maxArea = 0
        vis = [[False for _ in range(m)] for _ in range(n)]

        def bfs (r, c):
            cnt = 0
            q = deque()
            vis[r][c] = True
            q.append([r, c])

            while q:
                R, C = q.popleft()
                cnt += 1
                for i in range(len(dirs)):
                    nR = R + dirs[i][0]
                    nC = C + dirs[i][1]
                    if 0 <= nR < n and 0 <= nC < m and grid[nR][nC] == 1 and not vis[nR][nC]:
                        q.append([nR, nC])
                        vis[nR][nC] = True
            del q
            return cnt
        

        for i in range(n):
            for j in range(m):
                if grid[i][j] == 1 and not vis[i][j]:
                    maxArea = max(maxArea, bfs(i, j))
                    
        del vis
        del grid
        del dirs
        return maxArea